### PR TITLE
Add the `CITATION.cff` file to take advantage of GitHub's new support for the Citation File Format

### PR DESCRIPTION
**Pull request by storopoli**
This PR creates the initial draft for the CITATION.cff new file format.

Some resources for discussion:

* [CFF Documentation](https://citation-file-format.github.io/)
* [CFF documentation at GitHub](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/about-citation-files)

Note that with every Julia new release the `version` value should be updated. I briefly scanned for a GitHub Actions that could automatically do that but could not find (maybe since CFF are a new thing we should have new actions becoming available in the near future).

EDIT: I think we should leave the `.bib` also in the repo for now.

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Do we need the `version:` key? It would be easiest to just omit it.

If we are required to include the `version:` key, the value should match the value in the `VERSION` file: https://github.com/JuliaLang/julia/blob/master/VERSION

File: CITATION.cff [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+- family-names: "Bezanson"
+  given-names: "Jeff"
+- family-names: "Edelman"
+  given-names: "Alan"
+- family-names: "Karpinski"
+  given-names: "Stefan"
+- family-names: "Shah"
+  given-names: "Viral B"
+title: "Julia: A fresh approach to numerical computing"
+version: 1.6.2
```
Comment (DilumAluthge) [inline CITATION.cff:13]:
```suggestion
version: 1.8.0-DEV
```

This value should be the same as the value in the `VERSION` file: https://github.com/JuliaLang/julia/blob/master/VERSION

**Comment by storopoli** [comment]
Yes, we do otherwise the citation file is not valid/parsed by GitHub.

**Comment by DilumAluthge** [comment]
Brutal. That means that the release managers (@KristofferC and @ararslan) would need to remember to update the version number in two different places.

Probably the easiest is to have a tiny CI job that just checks that the `version:` in the `CITATION.cff` file exactly matches the contents of the `VERSION` file.

**Comment by DilumAluthge** [comment]
> EDIT: I think we should leave the `.bib` also in the repo for now.

Yeah I think that makes sense.

File: CITATION.cff (hunk 1 above)
Comment (ViralBShah) [reply CITATION.cff:13]: Also, we'll need to then change it for releases and such as well. What if we just don't include a version?

**Comment by ViralBShah** [comment]
Is it possible to get the rest of the bibliographic data into this format - like the journal and volume and such?

Also should we get rid of the CITATION.bib file, since this gives you an option to export a bib file? The only issue with that is that we refer to it in some places on discourse and such - but presumably we can update those links and post updated messages.

File: CITATION.cff (hunk 1 above)
Comment (DilumAluthge) [reply CITATION.cff:13]: See @storopoli's comment below - apparently GitHub will consider it invalid without the field.

Comment (DilumAluthge) [reply CITATION.cff:13]: > Also, we'll need to then change it for releases and such as well.

Yeah it's going to get really annoying really quickly.

Could we just say `version: 1`? It's technically correct. And then we don't need to update it for a while 😂

Comment (DilumAluthge) [reply CITATION.cff:13]: Or `version: v1`, or something.

**Comment by DilumAluthge** [comment]
> Also should we get rid of the CITATION.bib file, since this gives you an option to export a bib file? The only issue with that is that we refer to it in some places on discourse and such - but presumably we can update those links and post updated messages.

I was thinking keep it around. The contents of CITATION.bib are static, so can't hurt to keep it, right? At least for the short term, while we wait for the rest of the Julia ecosystem to start adopting this new standard.

**Comment by storopoli** [comment]
> Is it possible to get the rest of the bibliographic data into this format - like the journal and volume and such?
> 
> Also should we get rid of the CITATION.bib file, since this gives you an option to export a bib file? The only issue with that is that we refer to it in some places on discourse and such - but presumably we can update those links and post updated messages.

Yes, we can I will add the info.

**Comment by ViralBShah** [comment]
> Darn. That means that the release managers (@KristofferC and @ararslan) would need to remember to update the version number in two different places.
> 
> Probably the easiest is to have a tiny CI job that just checks that the `version:` in the `CITATION.cff` file exactly matches the contents of the `VERSION` file.

Is it possible to have a CI job just for releases? Running a tiny CI to check this on every commit feels a bit wasteful.

**Comment by DilumAluthge** [comment]
> > Darn. That means that the release managers (@KristofferC and @ararslan) would need to remember to update the version number in two different places.
> > Probably the easiest is to have a tiny CI job that just checks that the `version:` in the `CITATION.cff` file exactly matches the contents of the `VERSION` file.
> 
> Is it possible to have a CI job just for releases? Running a tiny CI to check this on every commit feels a bit wasteful.

Yeah, with Buildkite, we can have different CI for release branches vs master.

File: CITATION.cff (hunk 1 above)
Comment (vchuravy) [reply CITATION.cff:13]: @arfon can we trick the version field like that?

File: CITATION.cff [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      - family-names: "Karpinski"
+        given-names: "Stefan"
+      - family-names: "Shah"
+        name-particle: "B"
+        given-names: "Viral"
+    doi: "10.1137/141000671"
+    journal: "SIAM Review"
+    voume: 59
+    number: 1
+    pages: 13
```
Comment (storopoli) [inline CITATION.cff:37]: According to the [CFF documentation](https://github.com/citation-file-format/citation-file-format/blob/main/README.md#reference-keys) we don't have anything like the `pages={65--98}`:



`pages` | Integer | The number of pages of the work
-- | -- | --

Comment (ViralBShah) [reply CITATION.cff:37]: Can we put `65-98` in pages? Or will it freak out?

Comment (storopoli) [reply CITATION.cff:37]: I don't know... 😂 could not find a `.cff` parser online and the only thing close to that was the GitHub repository (after adding the `.cff` file)

Comment (ViralBShah) [reply CITATION.cff:37]: It's odd that everything has to be mentioned twice.

Comment (DilumAluthge) [reply CITATION.cff:37]: > Can we put `65-98` in pages? Or will it freak out?


Here's the schema, I think: https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml

It says that `pages` is an `int`.

File: CITATION.cff [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  given-names: "Jeff"
+- family-names: "Edelman"
+  given-names: "Alan"
+- family-names: "Karpinski"
+  given-names: "Stefan"
+- family-names: "Shah"
+  name-particle: "B"
+  given-names: "Viral"
+title: "Julia: A fresh approach to numerical computing"
+version: 1.8.0-DEV
```
Comment (DilumAluthge) [inline CITATION.cff:14]:
```suggestion
version: v1
```

**Comment by DilumAluthge** [comment]
@storopoli Where did you see that `version` is required?

The CFF schema (https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml) says that `version` is not required.

File: CITATION.cff (hunk 2 above)
Comment (DilumAluthge) [reply CITATION.cff:37]: I think `start` and `end` are the start and end pages, respectively.

**Comment by storopoli** [comment]
> @storopoli Where did you see that `version` is required?
> 
> The CFF schema ([https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml](https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml?rgh-link-date=2021-07-29T12%3A55%3A19Z)) says that `version` is not required.

Here: https://github.com/citation-file-format/citation-file-format/blob/main/README.md#example

> If you want to make your software easily citable, you can put a file called CITATION.cff in the root of your repository. This file should provide at least the minimally necessary metadata to cite your software.


And also I did this for some repos earlier today and GitHub did not parsed if there wasn't a version in the `.cff` file (https://github.com/JuliaDataScience/JuliaDataScience/pull/39/commits/544f615800299699761dbed800695ba5447d9fd2)

**Comment by DilumAluthge** [comment]
> > @storopoli Where did you see that `version` is required?
> > The CFF schema ([https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml](https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml?rgh-link-date=2021-07-29T12%3A55%3A19Z)) says that `version` is not required.
> 
> Here: https://github.com/citation-file-format/citation-file-format/blob/main/README.md#example
> 
> > If you want to make your software easily citable, you can put a file called CITATION.cff in the root of your repository. This file should provide at least the minimally necessary metadata to cite your software.
> 
> And also I did this for some repos earlier today and GitHub did not parsed if there wasn't a version in the `.cff` file ([JuliaDataScience/JuliaDataScience@544f615](https://github.com/JuliaDataScience/JuliaDataScience/commit/544f615800299699761dbed800695ba5447d9fd2))

Interesting. It seems to violate the schema, but 🤷

**Comment by DilumAluthge** [comment]
I think the `v1` solution will probably be fine. The alternative (manually updating it every time) seems unpleasant.

**Comment by storopoli** [comment]
> > > @storopoli Where did you see that `version` is required?
> > > The CFF schema ([https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml](https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml?rgh-link-date=2021-07-29T13%3A00%3A52Z)) says that `version` is not required.
> > 
> > 
> > Here: [https://github.com/citation-file-format/citation-file-format/blob/main/README.md#example](https://github.com/citation-file-format/citation-file-format/blob/main/README.md?rgh-link-date=2021-07-29T13%3A00%3A52Z#example)
> > > If you want to make your software easily citable, you can put a file called CITATION.cff in the root of your repository. This file should provide at least the minimally necessary metadata to cite your software.
> > 
> > 
> > And also I did this for some repos earlier today and GitHub did not parsed if there wasn't a version in the `.cff` file ([JuliaDataScience/JuliaDataScience@544f615](https://github.com/JuliaDataScience/JuliaDataScience/commit/544f615800299699761dbed800695ba5447d9fd2))
> 
> Interesting. It seems to violate the schema, but 🤷

No it doesn't. It has a `required: True`. There are several `version` in the schema.

See this one: https://github.com/citation-file-format/citation-file-format/blob/main/schema.yaml#L115-L117

**Comment by DilumAluthge** [comment]
🤦 You're exactly right, I was looking at the wrong `version`. Thanks!

**Comment by DilumAluthge** [comment]
Since there is a schema, presumably we can validate our `.cff` against the schema? E.g. with https://www.npmjs.com/package/yaml-schema-validator?

File: CITATION.cff [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  given-names: "Stefan"
+- family-names: "Shah"
+  name-particle: "B"
+  given-names: "Viral"
+title: "Julia: A fresh approach to numerical computing"
+version: "v1"
+license: "MIT"
+doi: "10.1137/141000671"
+date-released: 2017-02-07
+url: "https://github.com/JuliaLang/julia"
```
Comment (DilumAluthge) [inline CITATION.cff:18]: Which of the following makes the most sense for `url`?
1. `https://github.com/JuliaLang/julia`
2. `https://julialang.org`
3. `https://epubs.siam.org/doi/10.1137/141000671`
3. `https://dx.doi.org/10.1137/141000671` (which redirects to number 3 above)

File: CITATION.cff [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        given-names: "Viral"
+    doi: "10.1137/141000671"
+    journal: "SIAM Review"
+    volume: 59
+    number: 1
+    pages: 33
+    start: 65
+    end: 98
+    year: 2017
+    url: "https://epubs.siam.org/doi/10.1137/141000671"
```
Comment (DilumAluthge) [inline CITATION.cff:41]: Should this `url` be `https://epubs.siam.org/doi/10.1137/141000671` or `https://dx.doi.org/10.1137/141000671`?

The latter redirects to the former, and in theory the latter is more "future-proof".

File: CITATION.cff (hunk 4 above)
Comment (ViralBShah) [reply CITATION.cff:18]: Good point. I think we should use julialang.org. This is referring to Julia the system, not Julia the codebase.

File: CITATION.cff (hunk 5 above)
Comment (ViralBShah) [reply CITATION.cff:41]: Yes, preferable to use the doi.org URL.

File: CITATION.cff (hunk 4 above)
Comment (DilumAluthge) [reply CITATION.cff:18]:
```suggestion
url: "https://julialang.org"
```

File: CITATION.cff (hunk 5 above)
Comment (DilumAluthge) [reply CITATION.cff:41]:
```suggestion
    url: "https://dx.doi.org/10.1137/141000671"
```

File: CITATION.cff [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    title: "Julia: A fresh approach to numerical computing"
+    authors:
+      - family-names: "Bezanson"
+        given-names: "Jeff"
+      - family-names: "Edelman"
+        given-names: "Alan"
+      - family-names: "Karpinski"
+        given-names: "Stefan"
+      - family-names: "Shah"
+        name-particle: "B"
```
Comment (DilumAluthge) [inline CITATION.cff:31]: From https://github.com/citation-file-format/citation-file-format/blob/main/README.md#exemplary-uses-2, is "B" a name particle, or is it part of the given names?

@ViralBShah

**Comment by vtjnash** [comment]
If we decide to change the version, there is a release-checklist in the Makefile where that step would go

File: CITATION.cff (hunk 6 above)
Comment (ViralBShah) [reply CITATION.cff:31]: Yikes I don't understand so much about names. It is my middle name (first initial for my father's name) - which I use due to so many Viral Shahs out in the wild. I would say that becomes part of my given name.

Comment (storopoli) [reply CITATION.cff:31]: The example for `name-particle` in the documentation is:

```yaml
- family-names: "Beethoven"
  name-particle: "van"
  given-names: "Wolfgang"
```

Comment (ViralBShah) [reply CITATION.cff:31]: I looked at the docs - it is not a name-participle in my case. It really should be "Viral B." as my given name. Or if it is all confusing, just "Viral" as the given name.

Comment (storopoli) [reply CITATION.cff:31]: Let's do `Viral B.`

**Comment by ViralBShah** [comment]
I say we merge this - and we can tweak further later on as necessary.

**Comment by ViralBShah** [comment]
The generated bibtex is terrible.

**Comment by storopoli** [comment]
At least we have the dropdown to cite this repo:

image

**Comment by arfon** [comment]
> The github generated bibtex is terrible. I think we'll have to keep CITATION.bib for a while!

Feel free to open an issue on https://github.com/citation-file-format/ruby-cff reporting any issues with the bibtex 😁

**Comment by arfon** [comment]
> The github generated bibtex is terrible. I think we'll have to keep CITATION.bib for a while!

Feel free to open an issue on https://github.com/citation-file-format/ruby-cff reporting any issues with the generated bibtex output.

**Comment by ViralBShah** [comment]
The main thing is all the other metadata about the article that exists in our CITATION.cff file (like volume, pages, etc.) doesn't get incorporated into the generated BibTeX. Also it makes it an `@misc` bibtex entry, but we really want `@article`.

**Comment by arfon** [comment]
Got it, thanks! CFF v1.2.0 will be introducing the concept of a `preferred-citation` which can be an article rather than a (`@misc`) citation snippet for the _software_.

**Comment by StefanKarpinski** [comment]
To clarify why `@misc` is inappropriate for us, there is an actual academic publication that should be cited when citing this project, rather than directly citing the repository or the software. I suspect this is the case for many other Julia packages as well, and it is probably not uncommon in other language ecosystems.

**Pull request merged**